dh.c

```diff
@@ -283,6 +288,103 @@ dh_pub_is_valid(const DH *dh, const BIGNUM *dh_pub)
 int
 dh_gen_key(DH *dh, int need)
 {
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
```

Comment:
If you are just going to switch over to EVP_PKEY, this should be possible without wholesale code duplication for OpenSSL 3.0 -  The EVP_pkey interface isn't significantly different between 1.1.1 and 3.0.

Reply:
I think as a first step you should separate out all the OpenSSL 3.0 stuff from this, and make a PR that *ONLY* converts to the EVP_PKEY API with the LibreSSL (1.1.1) api.